R/Gargle-class.R

```diff
@@ -99,115 +110,33 @@ Gargle2.0 <- R6::R6Class("Gargle2.0", inherit = httr::Token2.0, list(
     }
   },
   print = function(...) {
-    cat("<Token (via gargle)>\n", sep = "")
-    # print(self$endpoint) ## this is TMI IMO
-    # also, it's boring --> always google for us
-    cat("  <oauth_endpoint> google\n", sep = "")
-    # print(self$app) ## this is TMI IMO
-    cat("       <oauth_app> ", self$app$appname, "\n", sep = "")
-    cat("           <email> ", self$email, "\n", sep = "")
-    ## TODO(jennybc) turn this into a helper or sg
-    scopes <- gsub("/$", "", gsub("(.*)/(.+$)", "...\\2", self$params$scope))
-    cat(
-      "          <scopes> ",
-      paste0(scopes, collapse = ", "),
-      "\n", sep = ""
-    )
-    cat(
-      "     <credentials> ",
-      paste0(names(self$credentials), collapse = ", "),
-      "\n", sep = ""
-    )
-    cat(
-      "                   (expires in ",
-      ceiling(self$credentials$expires_in / 60),
-      " mins)", "\n", sep = ""
-    )
-    cat("---\n")
+    cat_line("<Token (via gargle)>")
+    cat_line("  <oauth_endpoint> google")
+    cat_line("       <oauth_app> ", self$app$appname)
+    cat_line("           <email> ", self$email)
+    cat_line("          <scopes> ", commapse(base_scope(self$params$scope)))
+    cat_line("     <credentials> ", commapse(names(self$credentials)))
+    cat_line("---")
   },
   hash = function() {
-    # endpoint = which site = always google for us
-    # app = client identification = often tidyverse (or gargle) app
-    # params = scope <- this truly varies across client packages
-    msg <- serialize(list(
-      self$endpoint,
-      self$app,
-      normalize_scopes(self$params$scope)
-    ), NULL)
-
-    # for compatibility with digest::digest()
-    msg <- paste(openssl::md5(msg[-(1:14)]), collapse = "")
-
-    # append the email
-    paste(msg, self$email, sep = "-")
+    msg <- rhash(list(self$app, self$params$scope))
+    paste(msg, self$email, sep = "_")
```

Comment:
I no longer include endpoint here, ok? Why not? because it's always google here. Why I even ask? Because httr includes endpoint in the hash (but has no notion of email).

Reply:
The hash also includes normalized scopes and email. I think it's still a pretty handy way to look for an exact match.